homework-21

OneAndOnlyAdam opened the pull request:
Error for BaseTest line 45 suggested I change it from: wait = new WebDriverWait (driver, Duration. ofSeconds(10))

To:  wait = (WebDriver) new WebDriverWait(driver,Duration.ofSeconds(10)); Can you tell me why?  When during class the instructor had it like the top example.

In both the BaseTest class there's an issue on lines 60,64,69 and Homework21 lines 26,30,36 with the 'until'. rename or change to get is suggested.  What is wrong with it???

OneAndOnlyAdam committed:
homework-21

Error for BaseTest line 45 suggested I change it from:
wait = new WebDriverWait (driver, Duration. ofSeconds(10))

To:  wait = (WebDriver) new WebDriverWait(driver,Duration.ofSeconds(10));
Can you tell me why?  When during class the instructor had it like the top example.

In both the BaseTest class there's an issue on lines 60,64,69 and Homework21 lines 26,30,36 with the 'until'. rename or change to get is suggested.  What is wrong with it???

Berichicko commented:
> Error for BaseTest line 45 suggested I change it from: wait = new WebDriverWait (driver, Duration. ofSeconds(10))
> 
> To: wait = (WebDriver) new WebDriverWait(driver,Duration.ofSeconds(10)); Can you tell me why? When during class the instructor had it like the top example.
> 
> In both the BaseTest class there's an issue on lines 60,64,69 and Homework21 lines 26,30,36 with the 'until'. rename or change to get is suggested. What is wrong with it???

Hi Adam

You got this problem, please correct it. 

![image](https://github.com/Test-Pro-Bootcamp/A51-QA-Automation-Framework/assets/104761482/d2014051-7437-4e78-9b0b-0d6450d51992)

HINT

_The issue is that the until method is not a property of the WebDriver class. Instead, it is a method of the WebDriverWait class, which is used to wait for a certain condition to be met before proceeding with the execution of the test. To fix the issue, you need to use the until method on an instance of the WebDriverWait class, not the WebDriver class._ 

 `  wait = new WebDriverWait(driver, Duration.ofSeconds(10));`

```
public void provideEmail(String email){
    WebElement emailField = wait.until(ExpectedConditions.visibilityOfElementLocated(By.cssSelector("input[type='email']")));
    emailField.sendKeys(email);
}
```

_For example_

```
import io.github.bonigarcia.wdm.WebDriverManager;
import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.chrome.ChromeDriver;
import org.openqa.selenium.chrome.ChromeOptions;
import org.openqa.selenium.interactions.Actions;
import org.openqa.selenium.support.ui.ExpectedConditions;
import org.openqa.selenium.support.ui.WebDriverWait;
import org.testng.annotations.AfterMethod;
import org.testng.annotations.BeforeMethod;
import org.testng.annotations.BeforeSuite;
import org.testng.annotations.Parameters;

import java.time.Duration;
import java.util.UUID;

public class BaseTest {

    public WebDriver driver = null;
    public WebDriverWait wait = null;
    public String url = "https://qa.koel.app";
    public Actions actions = null;

    @BeforeSuite
    static void setupClass() {
        WebDriverManager.chromedriver().setup();
    }

    @BeforeMethod
    @Parameters({"BaseUrl"})
    public void launchBrowser(String BaseUrl) throws InterruptedException{
        // Added ChromeOptions argument to fix websocket error
        ChromeOptions options = new ChromeOptions();
        options.addArguments("--remote-allow-origins=*");

        // Open Chrome Browser
        driver = new ChromeDriver(options);
        actions = new Actions(driver);
        driver.manage().timeouts().implicitlyWait(Duration.ofSeconds(10));
        driver.manage().window().maximize();
        wait = new WebDriverWait(driver, Duration.ofSeconds(10));
        url = BaseUrl;
        navigateToPage();
    }

    @AfterMethod
    public void closeBrowser(){
        driver.quit();
    }

    // Prerequisites
    public void navigateToPage(){
        driver.get(url);
    }

    public void provideEmail(String email){
        WebElement emailField = wait.until(ExpectedConditions.visibilityOfElementLocated(By.cssSelector("input[type='email']")));
        emailField.sendKeys(email);
    }

    public void providePassword(String password){
        WebElement passwordField = wait.until(ExpectedConditions.visibilityOfElementLocated(By.cssSelector("input[type='password']")));
        passwordField.clear();
        passwordField.sendKeys(password);
    }

    public void clickSubmit(){
        WebElement submit = wait.until(ExpectedConditions.visibilityOfElementLocated(By.cssSelector("button[type='submit']")));
        submit.click();
    }
}
```

OneAndOnlyAdam committed:
Update BaseTest.java

I'm pretty sure that I fixed the errors I had.  Now I'm getting the same type of error that I was getting for Homework22.

This is the error:

org.gradle.api.internal.tasks.testing.TestSuiteExecutionException: Could not complete execution for Gradle Test Executor 2.

Same error type that Muhammad was getting in class last night as well.

Berichicko commented:
**HINT**

**Step 1**

Add  `<parameter name="BaseURL" value="https://qa.koel.app/"/>`

```
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE suite SYSTEM "http://testng.org/testng-1.0.dtd" >

<suite name="TestNG">
    <parameter name="BaseUrl" value="https://qa.koel.app/"/>
    <test name="Example Test" preserve-order="false">
        <classes>
            <class name="Homework21"/>
            <class name="BaseTest"/>
        </classes>

    </test>
</suite>
```

**Step 2**

 ` Actions actions = null;`
 
 
![image](https://github.com/Test-Pro-Bootcamp/A51-QA-Automation-Framework/assets/104761482/24545537-cca3-4373-8f0e-313ad7c2a3a3)

**Step 3**

Add  `doubleClickPlaylist();`

```
import org.openqa.selenium.By;
import org.openqa.selenium.Keys;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.support.ui.ExpectedConditions;
import org.testng.Assert;
import org.testng.annotations.Test;

public class Homework21 extends BaseTest {

    // Define a variable to store the new playlist name
    String newPlaylistName = "Abracadabra";

    // This is a TestNG test method to rename a playlist
    @Test
    public void renamePlaylist() {
        // Login
        provideEmail("adam.johnson@testpro.io");
        providePassword("1Te$t$tudent");
        clickSubmit();

        // Double click on the playlist to edit its name
        doubleClickPlaylist();

        // Enter the new playlist name
        enterNewPlaylistName();

        // Assertion to check if the new playlist exists
        Assert.assertTrue(doesPlaylistExist());
    }

    // Method to double click on a playlist
    public void doubleClickPlaylist() {
        // Locate the playlist element and perform a double click action
        WebElement playlistElement = wait.until(ExpectedConditions.visibilityOfElementLocated(By.cssSelector(".playlist:nth-child(3)")));
        actions.doubleClick(playlistElement).perform();
    }

    // Method to enter a new playlist name
    public void enterNewPlaylistName() {
        // Locate the input field for the playlist name, clear its content, enter a new name, and press ENTER
        WebElement playlistInputField = wait.until(ExpectedConditions.visibilityOfElementLocated(By.cssSelector("[name='name']")));
        playlistInputField.sendKeys(Keys.chord(Keys.CONTROL, "a", Keys.BACK_SPACE)); // Select all and delete
        playlistInputField.sendKeys(newPlaylistName); // Enter the new playlist name
        playlistInputField.sendKeys(Keys.ENTER); // Press ENTER to save the new name
    }

    // Method to check if the playlist with the new name exists
    public boolean doesPlaylistExist() {
        // Locate the playlist element with the new name and check if it is displayed
        WebElement playlistElement = wait.until(ExpectedConditions.visibilityOfElementLocated(By.linkText("Abracadabra")));
        return playlistElement.isDisplayed();
    }
}
```
